pkg/resources/grant_role.go

```diff
@@ -0,0 +1,175 @@
+					}
+				default:
+					return nil, fmt.Errorf("invalid object type specified: %v, expected ROLE or USER", parts[1])
+				}
+
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+	}
+}
+
+// CreateGrantRole implements schema.CreateFunc.
+func CreateGrantRole(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
+	roleName := d.Get("role_name").(string)
+	roleIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(roleName)
+	// format of snowflakeResourceID is <role_identifier>|<object type>|<target_identifier>
+	var snowflakeResourceID string
+	if parentRoleName, ok := d.GetOk("parent_role_name"); ok && parentRoleName.(string) != "" {
+		parentRoleIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(parentRoleName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(roleIdentifier.FullyQualifiedName(), sdk.ObjectTypeRole.String(), parentRoleIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantRoleRequest(roleIdentifier, sdk.GrantRole{
+			Role: &parentRoleIdentifier,
+		})
+		if err := client.Roles.Grant(ctx, req); err != nil {
+			return err
+		}
+	} else if userName, ok := d.GetOk("user_name"); ok && userName.(string) != "" {
+		userIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(userName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(roleIdentifier.FullyQualifiedName(), sdk.ObjectTypeUser.String(), userIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantRoleRequest(roleIdentifier, sdk.GrantRole{
+			User: &userIdentifier,
+		})
+		if err := client.Roles.Grant(ctx, req); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("invalid role grant specified: %v", d)
+	}
+	d.SetId(snowflakeResourceID)
+	return ReadGrantRole(d, meta)
+}
+
+func ReadGrantRole(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	parts := strings.Split(d.Id(), helpers.IDDelimiter)
+	roleName := parts[0]
+	roleIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(roleName)
+	objectType := parts[1]
+	targetIdentifier := parts[2]
+	ctx := context.Background()
+	grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
+		Of: &sdk.ShowGrantsOf{
+			Role: roleIdentifier,
+		},
+	})
+	if err != nil {
+		log.Printf("[DEBUG] role (%s) not found", roleIdentifier.FullyQualifiedName())
+		d.SetId("")
+		return nil
+	}
+
+	var found bool
+	for _, grant := range grants {
+		if grant.GrantedTo == sdk.ObjectType(objectType) {
+			if grant.GranteeName.FullyQualifiedName() == targetIdentifier {
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		log.Printf("[DEBUG] role grant (%s) not found", d.Id())
+		d.SetId("")
+	}
+
+	return nil
+}
+
+func DeleteGrantRole(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	parts := strings.Split(d.Id(), helpers.IDDelimiter)
```

Comment:
we could have a safety check like in the importer func here (i.e. `if len(parts) != 3)`). But this is an optional addition.

Reply:
added check